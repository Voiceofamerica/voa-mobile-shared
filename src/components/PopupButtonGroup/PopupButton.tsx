
import * as React from 'react'

import { themed, ThemeProps, DEFAULT_THEME } from '../ThemeProvider'

import {
  button,
  content,
} from './PopupButtonGroup.scss'

export interface Props<T = {}> extends ThemeProps {
  onClick: () => void
  className?: string
  style?: React.CSSProperties
}

@themed
class PopupButton extends React.Component<Props> {
  render () {
    const { onClick = () => null, className = '', style = {}, theme = DEFAULT_THEME, children } = this.props
    const {
      popupButtonBorder: borderColor,
      popupButtonColor: color,
    } = theme

    return (
      <div
        onClick={() => onClick()}
        className={`${button} ${className}`}
        style={{ borderColor, color, ...style }}
      >
        <div className={content}>
          {children}
        </div>
      </div>
    )
  }
}

export default PopupButton
