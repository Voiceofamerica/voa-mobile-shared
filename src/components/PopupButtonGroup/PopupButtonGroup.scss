
@import "../../commonStyles/index";

$button-size: 40px;
$button-text-size: 30px;
$button-padding: $button-size - $button-text-size;
$transition-time: 0.3s;

.popupButtonGroup {
  display: flex;
  flex-direction: row;
  border-radius: 10px;
  overflow: hidden;
  position: fixed;
  left: 50vw;
  z-index: $menu-z-max + 1;
  user-select: none;
  opacity: 0;
  pointer-events: none;
  transform: translateX(-50%) translateY(20px);
  box-shadow: 0 0 rgba(0, 0, 0, 0.5);
  transition:
    $transition-time opacity,
    $transition-time transform,
    $transition-time box-shadow;

  &.show {
    pointer-events: auto;
    opacity: 1;
    transform: translateX(-50%);
    box-shadow: 0 3px 5px rgba(0, 0, 0, 0.5);
  }

  &.visible {
    transition:
      $transition-time opacity,
      $transition-time transform,
      $transition-time box-shadow,
      $transition-time top;
  }

  .button {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    font-size: $button-text-size;
    border-right: 1px solid;

    [dir='rtl'] & {
      border-right: none;
      border-left: 1px solid;
    }

    .content {
      @include touchable();

      display: flex;
      flex-direction: row;
      justify-content: center;
      padding: $button-padding;
    }

    &:last-of-type {
      border: none;
    }
  }
}
