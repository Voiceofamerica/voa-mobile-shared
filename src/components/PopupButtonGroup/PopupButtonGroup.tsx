
import * as React from 'react'

import { themed, ThemeProps, DEFAULT_THEME } from '../ThemeProvider'

import {
  popupButtonGroup,
  show,
} from './PopupButtonGroup.scss'

export interface Props extends ThemeProps {
  verticalPosition?: number
  show?: boolean
  className?: string
  style?: React.CSSProperties
}

@themed
class PopupButtonGroup extends React.Component<Props> {
  render () {
    const { verticalPosition = 0, show: isShown = false, className = '', style = {}, theme = DEFAULT_THEME, children } = this.props
    const {
      popupButtonBackground: background,
    } = theme

    const fullClassName = isShown
                        ? `${popupButtonGroup} ${show} ${className}`
                        : `${popupButtonGroup} ${className}`

    return (
      <div className={fullClassName} style={{ background, top: verticalPosition, ...style }}>
        {children}
      </div>
    )
  }
}

export default PopupButtonGroup
