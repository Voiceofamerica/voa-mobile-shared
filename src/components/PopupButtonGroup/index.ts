
export { default } from './PopupButtonGroup'
export { default as PopupButton } from './PopupButton'
