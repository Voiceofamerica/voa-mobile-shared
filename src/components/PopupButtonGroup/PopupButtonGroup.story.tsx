
import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import ThemeProvider, { BaseTheme, PopupButtonGroupTheme, DEFAULT_THEME } from '../ThemeProvider'
import SvgIcon from '../SvgIcon'

import PopupButton from './PopupButton'
import PopupButtonGroup, { Props as PopupButtonGroupProps } from './PopupButtonGroup'

interface DisplayState {
  top: number
  show: boolean
}

class PopupButtonDisplay extends React.Component<Partial<PopupButtonGroupProps>, DisplayState> {
  state: DisplayState = {
    top: 0,
    show: false,
  }

  render () {
    const { children, ...props } = this.props
    const { top, show } = this.state

    return (
      <div>
        <div
          style={{
            fontSize: 30,
            background: '#bbbbbb',
            color: '#666666',
            display: 'flex',
            justifyContent: 'center',
            alignItems: 'center',
            position: 'fixed',
            left: 0,
            right: 0,
            top: 0,
            bottom: 0,
            cursor: 'pointer',
          }}
          onClick={this.handleClick}
        >
          Click anywhere to show buttons
        </div>
        <PopupButtonGroup {...props} verticalPosition={top} show={show}>
          {children}
          <PopupButton onClick={this.cancel} style={{ fontSize: 15 }}>
            Cancel
          </PopupButton>
        </PopupButtonGroup>
      </div>
    )
  }

  private handleClick = (ev: React.MouseEvent<HTMLDivElement>) => {
    const top = ev.clientY + 10
    this.setState({ top, show: true })
  }

  private cancel = () => {
    this.setState({ show: false })
  }
}

const testBaseTheme: BaseTheme = {
  lightGrey: '#666666',
  black: '#DDDDFF',
  grey: '#AAAAAA',
}

const testLocalTheme: PopupButtonGroupTheme = {
  popupButtonBackground: '#003300',
  popupButtonColor: '#DDFFDD',
  popupButtonBorder: '#664499',
}

storiesOf(PopupButtonGroup.name, module)
  .add('default display', () => (
    <PopupButtonGroup verticalPosition={50} show>
      <PopupButton onClick={action('x button')}>
        <SvgIcon src='close' style={{ color: DEFAULT_THEME.red }} />
      </PopupButton>
      <PopupButton onClick={action('up button')}>
        <SvgIcon src='chevronUp' />
      </PopupButton>
      <PopupButton onClick={action('down button')}>
        <SvgIcon src='chevronDown' />
      </PopupButton>
    </PopupButtonGroup>
  ))
  .add('positionable display', () => (
    <PopupButtonDisplay>
      <PopupButton onClick={action('x button')}>
        <SvgIcon src='close' style={{ color: DEFAULT_THEME.red }} />
      </PopupButton>
      <PopupButton onClick={action('up button')}>
        <SvgIcon src='chevronUp' />
      </PopupButton>
      <PopupButton onClick={action('down button')}>
        <SvgIcon src='chevronDown' />
      </PopupButton>
    </PopupButtonDisplay>
  ))
  .add('base themed display', () => (
    <ThemeProvider value={testBaseTheme}>
      <PopupButtonGroup verticalPosition={50} show>
        <PopupButton onClick={action('x button')}>
          <SvgIcon src='close' style={{ color: DEFAULT_THEME.red }} />
        </PopupButton>
        <PopupButton onClick={action('up button')}>
          <SvgIcon src='chevronUp' />
        </PopupButton>
        <PopupButton onClick={action('down button')}>
          <SvgIcon src='chevronDown' />
        </PopupButton>
      </PopupButtonGroup>
    </ThemeProvider>
  ))
  .add('local themed display', () => (
    <ThemeProvider value={testLocalTheme}>
      <PopupButtonGroup verticalPosition={50} show>
        <PopupButton onClick={action('x button')}>
          <SvgIcon src='close' style={{ color: DEFAULT_THEME.red }} />
        </PopupButton>
        <PopupButton onClick={action('up button')}>
          <SvgIcon src='chevronUp' />
        </PopupButton>
        <PopupButton onClick={action('down button')}>
          <SvgIcon src='chevronDown' />
        </PopupButton>
      </PopupButtonGroup>
    </ThemeProvider>
  ))
