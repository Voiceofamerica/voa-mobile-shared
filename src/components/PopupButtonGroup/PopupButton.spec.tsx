
import * as React from 'react'
import { create } from 'react-test-renderer'
import { mount } from 'enzyme'

import PopupButton from './PopupButton'
import { button } from './PopupButtonGroup.scss'

describe(`<${PopupButton.name} />`, () => {
  describe('shapshots', () => {
    it('should render without children', () => {
      const element = create((
        <PopupButton onClick={jest.fn()} />
      ))

      expect(element).toMatchSnapshot()
    })
    it('should render with children', () => {
      const element = create((
        <PopupButton onClick={jest.fn()}>
          This should show up
        </PopupButton>
      ))

      expect(element).toMatchSnapshot()
    })
  })

  describe('onClick', () => {
    it('should be called when the button is pressed', () => {
      const onClick = jest.fn()

      const tree = mount((
        <PopupButton onClick={onClick} />
      ))

      tree.find(`.${button}`).simulate('click')

      expect(onClick).toHaveBeenCalledTimes(1)
    })
  })
})
