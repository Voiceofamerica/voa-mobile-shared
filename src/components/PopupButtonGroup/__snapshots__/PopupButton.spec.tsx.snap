// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PopupButton /> shapshots should render with children 1`] = `
<div
  className="button "
  onClick={[Function]}
  style={
    Object {
      "borderColor": "#B4BCC2",
      "color": "#000000",
    }
  }
>
  <div
    className="content"
  >
    This should show up
  </div>
</div>
`;

exports[`<PopupButton /> shapshots should render without children 1`] = `
<div
  className="button "
  onClick={[Function]}
  style={
    Object {
      "borderColor": "#B4BCC2",
      "color": "#000000",
    }
  }
>
  <div
    className="content"
  />
</div>
`;
